enhancements/ebpf-in-ocp.md

```diff
@@ -0,0 +1,211 @@
+---
+title: ebpf-in-ocp
+authors:
+  - "@danwinship"
+reviewers:
+  - "@TBD, for network observability"
+  - "@TBD, for other existing eBPF use case X"
+  - "@TBD, for other future eBPF use case Y"
+  - "@dave-tucker, for bpfd"
+approvers:
+  - TBD
+api-approvers:
+  - None
+creation-date: 2022-05-25
+last-updated: 2022-05-25
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also:
+replaces:
+superseded-by:
+---
+
+# Guidelines for the Use of eBPF in OCP
+
+## Summary
+
+OCP developers and customers alike would like to use eBPF in OCP
+clusters. At the present time, _some_ eBPF hooks are designed to be
+"shareable", but others are not (and some are shareable only when used
+in certain restricted ways). We need to create rules for how OCP will
+use eBPF and how we expect others to use (or not use it) to prevent
+conflicts later.
+
+## Motivation
+
+### User Stories
+
+#### Existing Use Case Number 1
+
+(Something about observability or tracing where people are already
+using eBPF in OCP.)
+
+#### Upcoming Network Observability Features
+
+(Something about upcoming network observability features.)
+
+#### Ensuring Network (and General) Supportability
+
+As an OCP developer or support engineer, I would like to be able to
+debug customer problems without having to worry that random
+"unsanctioned" eBPF is messing things up in inexplicable ways.
+
+#### Customer eBPF Bug Patching
+
+As an OCP customer, I want to work around an OCP/RHEL bug by using a
+`MachineConfig` (or something similar) to deploy a small eBPF program
+to every node in my cluster, without voiding my warranty.
+
+I might be OK with needing a support exception, since this is just to
+work around a bug that should eventually be fixed.
+
+#### Customer eBPF Feature Patching
+
+As an OCP customer, I want to add functionality to OCP/RHEL by using a
+`MachineConfig` (or something similar) to deploy a small (or
+not-so-small) eBPF program to every node in my cluster, without
+voiding my warranty.
+
+I might be OK with needing a support exception.
+
+(eg, concretely, there is a customer who wants to use eBPF to
+implement [OVS-style SLB
+Bonding](https://docs.openvswitch.org/en/latest/topics/bonding/#slb-bonding)
+on a kernel bond interface, using eBPF. The upstream Linux maintainer
+rejected a patch to implement it in the kernel driver directly, in
+part because it would be so easy to implement it with eBPF.)
+
+#### Cilium as a Supported Third-Party Network Plugin
+
+As an OCP customer, I want to deploy an OCP cluster using Cilium as a
+certified third-party network plugin, without worrying that any other
+OCP components are going to interfere with it (or vice versa).
+
```

Comment:
I feel like CI will be the best way to ensure this... should we have an "ebpf feature" CI workflow to ensure stable cooperation? Maybe a job running with cilium + a job running with any OCP specific ebpf features

Reply:
(obviously good documentation on what hooks we're using explicitly is important to) but yeah the problem of ebpf interoperability is definitely a complex one.

Reply:
We do have Cilium jobs in CI - https://prow.ci.openshift.org/?job=*cilium*. They however run just the conformance suite, no custom eBPF tests

Reply:
additionally there is pixie https://docs.px.dev/about-pixie/pixie-ebpf/